### PR TITLE
DropWizard 0.8 upgrade using GuiceIntoHK2Bridge

### PR DESCRIPTION
**Pull request by yunspace**
To address issue #37 I've updated the code to incoroprate [Guice/HK2 Bridge](https://hk2.java.net/guice-bridge/). A quick summary:

 * removed `jersey.spi.inject` references and related classes such as `GuiceComponentProviderFactory`, `JerseyContainerModule`
 * GuiceContainer is instantiated by injection instead of `new GuiceContainer()`
 * GuiceContainer calls HK2 bridge during init() 
 * wrote tests and always use `javax.inject.Inject` annotations.

I noticed there is pull request #45 which uses [jersey2-guice](https://github.com/Squarespace/jersey2-guice). It would be interesting to compare the two different approaches.

### CODE REVIEW

**Comment by yunspace** [comment]
I wrote some tests to cover `AutoConfig`. In the process of testing, found out that dropwizard 0.8.x Tasks require a mandatory name in the constructor. I resolved this by adding a `@Named` injection in the Task constructor. Reflected this in my test cases and updated README.md

**Comment by jontejj** [comment]
Would be great if the tests in this commit could be added to #45. I like the tests in this commit but I like the code better in #45 (mostly because it's less code)

**Comment by jontejj** [comment]
From their readme:
"Jersey1 worked great with Guice. For Jersey2, the team wrote their own DI framework from scratch (HK2) and littered it with global state, config files in META-INF/service, and other JavaEE-isms that make my skin crawl. In the year and a half that Jersey2 has been released, nobody has managed to make it play 100% nicely with Guice. The most valiant attempt uses reflection to punch values into private final fields in HK2's static globals. It will likely break in future point releases of Jersey.

If I sound bitter writing this, it's because I wasted a stupid amount of time getting it to work. By comparison, RESTEasy comes pre-baked with Guice integration that worked almost immediately."

**Comment by yunspace** [comment]
@jontejj under the hood the jersey2-guice library would be doing similar things as HK2 guice-bridge, ie linking ServiceLocator to Guice Injector etc. I suppose one difference would be that jersey2-guice is specifically written for Jersey2 where as the guice-bridge is more generic. 

Personally I don't like HK2 either. But would prefer using HK2 provided tools directly rather than going through 3rd party code, potentially less layers of support when issues come up. I agree with the gwizard author in that: `the jersey2-guice adapter (used by this module) may break with future Jersey point releases`. 

We've used HK2 guice-bridge for our Jersey2 GAE app in prod environment for the second half of 2014. We didn't come across any issues, which is why I'm quite confident in migrating it over to DW. Also reflected by the tests.

I'll upgrade this pull request to use 0.8.0-rc2 soon. If anyone wishes to go for a test run, the forked artifact can be accessed form our company bintray: https://bintray.com/trunkplatform/dropwizard-modules/trunkplatform%3Adropwizard-guice/view

**Comment by jontejj** [comment]
We'll definitely give that a go!

**Comment by jontejj** [comment]
I tried your 0.8.0-rc2-201501201620 release, During the initialization phase in dropwizard I get "No implementation for javax.ws.rs.core.HttpHeaders was bound." Is https://jersey.java.net/apidocs/1.8/contribs/jersey-guice/com/sun/jersey/guice/JerseyServletModule.html not ported somehow?

File: src/test/java/com/hubspot/dropwizard/guice/AutoConfigTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        assertThat(healthChecks).contains(new InjectedHealthCheck().getName());
+    }
+
+    @Test
+    public void shouldAddProviders() {
+        // when
+        autoConfig.run(environment, injector);
+
+        //then
+        Set<Class<?>> components = environment.jersey().getResourceConfig().getClasses();
```
Comment (jontejj) [inline src/test/java/com/hubspot/dropwizard/guice/AutoConfigTest.java]: Perhaps verify that the instance is creatable with guice?

**Comment by oillio** [comment]
I wrote the jersey2-guice version.
Nice job @yunspace, the test suite looks nice.

I think either solution will work fine.
One major advantage of jersey2-guice is that it handles Guice JIT bindings.

I also wrote the beginnings of a test suite.  It is located in my other PR (along with some other goodies I have been working on): https://github.com/HubSpot/dropwizard-guice/pull/46
The suite isn't as extensive or as clean as the one by @yunspace and focuses more on my new features.  Also, it is built as integration tests rather than unit tests.

**Comment by yunspace** [comment]
Thanks for the compliments @oillio, your jersey2-guice PR inspired me to get this one in. The lack of support for JIT bindings in guide-bridge is painful indeed, if jersey2-guice handles it that's pretty awesome.

Yes I agree either solution will work fine. Perhaps it's wise to offer the option of letting the user choose which implementation to use. Ie having 2 subprojects, similar to dropwizard's freemarker vs mustache modules. Of course we would need a thorough test suite to ensure consistent functionality across the two.

File: src/test/java/com/hubspot/dropwizard/guice/AutoConfigTest.java (hunk 1 above)
Comment (yunspace) [inline src/test/java/com/hubspot/dropwizard/guice/AutoConfigTest.java]: Do you mean to verify healthcheck can be created by guice and that the registered health check is also same as the instance created by guice? Can do, will take a look tmr.

**Comment by yunspace** [comment]
I'm closing this PR for now. Until I find time to add support for JIT bindings and various other features, I'll be leveraging the jersey2-guice implementation (https://github.com/HubSpot/dropwizard-guice/pull/45) which offers most of the functionalities my project currently need.

I've squashed the commits and did some clean up in my code base. It's available at https://github.com/yunspace/dropwizard-guice/tree/guice-bridge if anyone is interested.

**Pull request closed without merging**